### ISSUE TITLE
Bug report: Empty CalendarPanel under Linux

### ISSUE THREAD

**Issue by mcwolfmm** [LGoodDatePicker/LGoodDatePicker#38]
Sometimes the DatePicker shows an empty CalendarPanel which can not be hidden until you close the window.

``` java
public class DateTimePickerSettings {
    private static final Locale LOCALE = Locale.forLanguageTag("bg-BG");

    public static DatePickerSettings getDateSettings() {
        DatePickerSettings settings = new DatePickerSettings();
        settings.setLocale(LOCALE);
        settings.setFirstDayOfWeek(DayOfWeek.MONDAY);
        settings.setAllowKeyboardEditing(false);
        settings.setFormatForDatesBeforeCommonEra("dd-MM-yyyy");
        settings.setFormatForDatesCommonEra("dd-MM-yyyy");
        settings.setTranslationClear("Изчисти");
        settings.setTranslationToday("Днес");
        settings.setGapBeforeButtonPixels(1);

        return settings;
    }

    public static TimePickerSettings getTimeSettings() {
        TimePickerSettings settings = new TimePickerSettings(LOCALE);
        settings.setFormatForDisplayTime("HH:mm");
        settings.setAllowKeyboardEditing(false);
        settings.setDisplayToggleTimeMenuButton(true);
        settings.setDisplaySpinnerButtons(false);
        settings.setGapBeforeButtonPixels(1);

        return settings;
    }
}
```

``` java
checkInPicker = new DateTimePicker(DateTimePickerSettings.getDateSettings(), DateTimePickerSettings.getTimeSettings());
checkInPicker.getDatePicker().getComponentToggleCalendarButton().setText(null);
checkInPicker.getDatePicker().getComponentToggleCalendarButton().setIcon(ImageHelper.getImageIcon(ImageHelper.CALENDAR_16x16));
checkInPicker.getDatePicker().setDateToToday();
checkInPicker.getDatePicker().addDateChangeListener(this);
checkInPicker.getDatePicker().getSettings().setAllowEmptyDates(false);
```

``` bash
$ uname -a
Linux service 4.4.0-38-generic #57-Ubuntu SMP Tue Sep 6 15:42:33 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
$ java -version
java version "1.8.0_92"
Java(TM) SE Runtime Environment (build 1.8.0_92-b14)
Java HotSpot(TM) 64-Bit Server VM (build 25.92-b14, mixed mode)
$ javac -version
javac 1.8.0_92
```

![bug](https://cloud.githubusercontent.com/assets/6696551/19339661/665b1d84-912d-11e6-912a-db80e72dd7af.png)

**Comment by mcwolfmm** [comment]
I see this issue https://github.com/LGoodDatePicker/LGoodDatePicker/issues/30, and I foud the bug.

First [here](https://github.com/LGoodDatePicker/LGoodDatePicker/blob/master/Project/src/main/java/com/github/lgooddatepicker/components/DatePicker.java#L990) must by
```java
        this.popup = null;
```
and secondly in `openPopup()`  must check wheter popup is open and must close.
```java
    public void openPopup() {
        if(isPopupOpen()) {
            closePopup();
            return;
        }
		
        if (settings == null) {
            return;
        }
        // If the component is disabled, do nothing.
```

This solves the problem of remaining empty window.

The next problem is the logic by hiding the window at shifting the focus in `class CustomPopup`. Sometimes the focus shifts before the window appears, which hides it instantly. It is therefore necessary to put a flag to prevent this happening.
```java
    /**
     * topWindow, This is a reference to the top window of the component which initiated the popup.
     * This is used for registering and deregistering the window movement listener with the
     * topWindow. If the top window moves, the popup is closed. This is set to null in the hide()
     * function
     */
    private Window topWindow;

    private boolean enableHideAction = false;
...

    // Create the display window.
    displayWindow = new JWindow(topWindow);
    displayWindow.addWindowListener(new WindowAdapter() {
        @Override
        public void windowOpened(WindowEvent e) {
            setEnableHideAction(true);
        }
    });

...
    /**
     * windowLostFocus, Part of WindowFocusListener. Whenever the popup window loses focus, it will
     * be hidden.
     */
    @Override
    public void windowLostFocus(WindowEvent e) {
        if(isEnableHideAction()) {
            hide();
        }
        else {
            e.getWindow().requestFocus();
        }
    }
	
    public void setEnableHideAction(boolean enabled) {
        enableHideAction = enabled;
    }

    public boolean isEnableHideAction() {
        return enableHideAction;
    }
```

**Comment by BlakeTNC** [comment]
Fixed with pull request #39.

**Comment by BlakeTNC** [comment]
The fix for this issue has been added to Release 8.2.0. LGoodDatePicker Version 8.2.0 has been published in the [Release Section](https://github.com/LGoodDatePicker/LGoodDatePicker/releases), as an [LGoodDatePicker Backport](https://github.com/LGoodDatePicker/LGoodDatePicker/releases/tag/Backports), and on [Maven Central](http://search.maven.org/#search%7Cga%7C1%7Ca%3A%22LGoodDatePicker%22).

@mcwolfmm Thank you for discovering this bug and developing the fix!

**Comment by kulik** [comment]
Hello @mcwolfmm Have Issue with this fix. wouldn't you mind to skype.its impossible to pick any date in case focus lost before mouse click action

`… [AWT-EventQueue-0] INFO com.github.lgooddatepicker.components.DatePicker - toggleCalendarButton --pressed
[AWT-EventQueue-0] INFO com.github.lgooddatepicker.components.DatePicker - togglePopup 
[AWT-EventQueue-0] INFO com.github.lgooddatepicker.zinternaltools.CustomPopup - com/github/lgooddatepicker/zinternaltools/CustomPopup.java:263
[AWT-EventQueue-0] INFO com.github.lgooddatepicker.components.CalendarPanel - mouse  dateLabel click`

So com/github/lgooddatepicker/zinternaltools/CustomPopup.java:263 called before mouse listener on the label
uname -a
Linux kulik-N550JK 3.13.0-101-generic #148-Ubuntu SMP Thu Oct 20 22:08:32 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

➜  src git:(master) ✗ uname -a
Linux kulik-N550JK 3.13.0-101-generic #148-Ubuntu SMP Thu Oct 20 22:08:32 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
➜  src git:(master) ✗ java -version
java version "1.8.0_111"
Java(TM) SE Runtime Environment (build 1.8.0_111-b14)
Java HotSpot(TM) 64-Bit Server VM (build 25.111-b14, mixed mode)

Unity graphical manager

Could you please mention way to reproduce your origin issue?

**Comment by BlakeTNC** [comment]
Hello @mcwolfmm and @kulik,

I've been working on this issue using a virtual machine with Linux Ubuntu 16.04. I believe I have authored a fix that will work for both of your versions of Linux. I need to request if each of you can download the latest code from the repository, and make sure everything is working properly on your machines. Or you could just run this demo jar and tell me if everything is working regarding focus lost events and selecting dates and times. 
[LGoodDatePicker-8.2.3-executable-demo - experimental.jar.zip](https://github.com/LGoodDatePicker/LGoodDatePicker/files/643444/LGoodDatePicker-8.2.3-executable-demo.-.experimental.jar.zip)

Thank you. 
Blake

**Comment by BlakeTNC** [comment]
Update: @kulik has told me that the fix is working for him. 
-Blake

**Comment by mcwolfmm** [comment]
hi,
this fix work fine and for me.

**Comment by BlakeTNC** [comment]
Thanks all for your help in making sure that this challenging bug was fully resolved. The new fix has been released in [LGoodDatePicker 8.3.0](https://github.com/LGoodDatePicker/LGoodDatePicker/releases), and on [Maven Central](http://search.maven.org/#search%7Cga%7C1%7Ca%3A%22LGoodDatePicker%22).
-Blake

**Comment by cchrisx31** [comment]
The bug is live on Linux Ubuntu 18.04
